### PR TITLE
*: update toolchain to 2022-07-31

### PR DESCRIPTION
**Pull request by tabokie**
### What is changed and how it works?

Issue Number: Ref #13008

What's Changed:

As a follow-up on #13009, enable new lint `assertions_on_result_states` to deny `assert!(r.is_ok())` pattern in the codebase.

```commit-message
```

### Check List 

Tests 

- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/concurrency_manager/src/lib.rs [hunk 1]
```diff
@@ -135,9 +135,9 @@ mod tests {
     #[tokio::test]
     async fn test_lock_keys_order() {
         let concurrency_manager = ConcurrencyManager::new(1.into());
-        let keys: Vec<_> = [b"c", b"a", b"b"]
-            .iter()
-            .map(|k| Key::from_raw(*k))
+        let keys: Vec<_> = vec![b"c", b"a", b"b"]
+            .into_iter()
+            .map(|k| Key::from_raw(k))
```
Comment (sticnarf) [inline components/concurrency_manager/src/lib.rs:140]: Umm, this change introduces an additional allocation for `vec![]` but brings no benefit.

As it's in a test, it's okay to just make clippy happy. But it's curious to know what does clippy complain about the old code?

**Review by sticnarf — COMMENTED** [review]
Please help upgrade `pin-project` to fix the warning raised by CI.

File: components/concurrency_manager/src/lib.rs (hunk 1 above)
Comment (tabokie) [reply components/concurrency_manager/src/lib.rs:140]: Check [this](https://github.com/rust-lang/rust-clippy/issues/9101).

File: components/engine_rocks/src/engine.rs [hunk 2]
```diff
@@ -250,7 +250,7 @@ mod tests {
         engine.put_cf(cf, b"k1", b"v2").unwrap();
 
         assert_eq!(&*engine.get_value(b"k1").unwrap().unwrap(), b"v1");
-        assert!(engine.get_value_cf("foo", b"k1").is_err());
+        engine.get_value_cf("foo", b"k1").unwrap_err();
```
Comment (Connor1996) [inline components/engine_rocks/src/engine.rs:253]: why

File: components/concurrency_manager/src/lib.rs (hunk 1 above)
Comment (sticnarf) [reply components/concurrency_manager/src/lib.rs:140]: Let's use `[b"c", b"a", b"b"].iter().copied()` just like other cases in this PR.

**Review by sticnarf — COMMENTED** [review]
Note that https://github.com/rust-lang/rust/pull/95685 will reopen https://github.com/tikv/tikv/issues/10906. As we are using frame pointers for the built-in profiling, the toolchain update will not affect profiling. But it may potentially make tikv not debuggable by gdb again (reported by https://github.com/tikv/tikv/issues/12091)

**Comment by sticnarf** [comment]
The rest lgtm

**Review by sticnarf — APPROVED** [review]

File: src/storage/types.rs [hunk 3]
```diff
@@ -80,7 +80,7 @@ impl MvccInfo {
 }
 
 /// Represents the status of a transaction.
-#[derive(PartialEq, Debug)]
+#[derive(PartialEq, Eq, Debug)]
```
Comment (BusyJay) [inline src/storage/types.rs:83]: Why need this?

Comment (tabokie) [reply src/storage/types.rs:83]: It’s the derive_partial_eq_without_eq lint.

Comment (BusyJay) [reply src/storage/types.rs:83]: Interesting. Will derive useless trait slow down compilation?

Comment (tabokie) [reply src/storage/types.rs:83]: I think it would: https://github.com/rust-lang/rust/issues/80118, I'll see if I can disable it.

Comment (tabokie) [reply src/storage/types.rs]: Updated and removed a bunch of unneeded `Eq`s.

File: components/raftstore/src/coprocessor/region_info_accessor.rs [hunk 4]
```diff
@@ -1166,8 +1166,8 @@ mod tests {
         ];
 
         for index in indices {
-            for order in orders {
-                test_split_impl(*index, *order);
+            for &order in orders {
```
Comment (BusyJay) [inline components/raftstore/src/coprocessor/region_info_accessor.rs:1169]: `*order` is a prefered style than `&order` matching.

File: scripts/clippy [hunk 5]
```diff
@@ -15,7 +15,12 @@ if [[ -n "$SHELL_DEBUG" ]] ; then
   set -x
 fi
 
-CLIPPY_LINTS=(-A clippy::module_inception  \
+# Notes:
+# - Enables `significant_drop_in_scrutinee` after
+#   https://github.com/rust-lang/rust-clippy/issues/8963 is fixed.
+# - `derive_partial_eq_without_eq` has compilation overhead.
+CLIPPY_LINTS=(
```
[the remark this answers is not in the document]
Comment (tabokie) [reply scripts/clippy:22]: It's only to keep diff short when adding a new lint in the future.

File: components/raftstore/src/coprocessor/region_info_accessor.rs (hunk 4 above)
Comment (tabokie) [reply components/raftstore/src/coprocessor/region_info_accessor.rs:1169]: I prefer this than putting a `[allow(explicit_auto_deref)]`.

Comment (BusyJay) [reply components/raftstore/src/coprocessor/region_info_accessor.rs:1169]: Then it's a false alarm from clippy as "deref which would be done by auto-deref" is false in this case. You can get around the problem by `order.as_slice()` or give orders a explicit type `orders: &[&[usize]]`.

File: components/engine_traits/src/cf_defs.rs [hunk 6]
```diff
@@ -14,11 +14,5 @@ pub fn name_to_cf(name: &str) -> Option<CfName> {
...
         return Some(CF_DEFAULT);
     }
-    for c in ALL_CFS {
-        if name == *c {
-            return Some(c);
-        }
-    }
-
-    None
+    ALL_CFS.iter().find(|&c| name == *c).copied()
```
Comment (BusyJay) [inline components/engine_traits/src/cf_defs.rs:17]: `copied` can also be used in this case.

**Review by BusyJay — APPROVED** [review]
Rest LGTM

File: components/engine_traits/src/cf_defs.rs (hunk 6 above)
Comment (tabokie) [reply components/engine_traits/src/cf_defs.rs:17]: Which case are you referring to? This LoC already uses `copied`.

Comment (sticnarf) [reply components/engine_traits/src/cf_defs.rs:17]: I think what BusyJay expects is like `ALL_CFS.iter().copied().find(|c| name == *c)`, so there is no reference matching.

Comment (tabokie) [reply components/engine_traits/src/cf_defs.rs:17]: Oh, right.

**Pull request merged**